sklearn/gaussian_process/gpr.py

```diff
@@ -0,0 +1,417 @@
+
+        Parameters
+        ----------
+        X : array-like, shape = (n_samples_X, n_features)
+            Query points where the GP samples are evaluated
+
+        n_samples : int, default: 1
+            The number of samples drawn from the Gaussian process
+
+        random_state: RandomState or an int seed (0 by default)
+            A random number generator instance
+
+        Returns
+        -------
+        y_samples : array, shape = (n_samples_X, [n_output_dims], n_samples)
+            Values of n_samples samples drawn from Gaussian process and
+            evaluated at query points.
+        """
+        rng = check_random_state(random_state)
+
+        y_mean, y_cov = self.predict(X, return_cov=True)
+        if y_mean.ndim == 1:
+            y_samples = rng.multivariate_normal(y_mean, y_cov, n_samples).T
+        else:
+            y_samples = \
+                [rng.multivariate_normal(y_mean[:, i], y_cov,
+                                         n_samples).T[:, np.newaxis]
+                 for i in range(y_mean.shape[1])]
+            y_samples = np.hstack(y_samples)
+        return y_samples
+
+    def log_marginal_likelihood(self, theta, eval_gradient=False):
+        """Returns log-marginal likelihood of theta for training data.
+
+        Parameters
+        ----------
+        theta : array-like, shape = (n_kernel_params,)
+            Kernel hyperparameters for which the log-marginal likelihood is
+            evaluated
+
+        eval_gradient : bool, default: False
+            If True, the gradient of the log-marginal likelihood with respect
+            to the kernel hyperparameters at position theta is returned
+            additionally.
+
+        Returns
+        -------
+        log_likelihood : float
+            Log-marginal likelihood of theta for training data.
+
+        log_likelihood_gradient : array, shape = (n_kernel_params,), optional
+            Gradient of the log-marginal likelihood with respect to the kernel
+            hyperparameters at position theta.
+            Only returned when eval_gradient is True.
+        """
+        kernel = self.kernel_.clone_with_theta(theta)
+
+        if eval_gradient:
+            K, K_gradient = kernel(self.X_train_, eval_gradient=True)
+        else:
+            K = kernel(self.X_train_)
+
+        K[np.diag_indices_from(K)] += self.sigma_squared_n
+        try:
+            L = cholesky(K, lower=True)  # Line 2
+        except np.linalg.LinAlgError:
+            return (-np.inf, np.zeros_like(theta)) \
+                if eval_gradient else -np.inf
+
+        log_likelihood = 0
+        if eval_gradient:
+            log_likelihood_gradient = 0
+
+        # Iterate over output dimensions of self.y_train_
+        y_fit = self.y_train_
+        if y_fit.ndim == 1:
+            y_fit = y_fit[:, np.newaxis]
+        for i in range(y_fit.shape[1]):
+            alpha = cho_solve((L, True), y_fit[:, i])  # Line 3
```

Comment:
In my scipy, version `0.16`, I can do `alpha = cho_solve((L, True), y_fit)`, which may be better than looping over targets. Is the loop here for some sort of backwards compatibility? If so, and the multitarget yields a speed up, there should probably be a scipy version test here.